File: ui-ngx/src/app/core/http/role.service.ts
///
/// Copyright © 2016-2026 The Thingsboard Authors
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { defaultHttpOptionsFromConfig, RequestConfig } from './http-utils';
import { PageData } from '@shared/models/page/page-data';
import { PageLink } from '@shared/models/page/page-link';

export interface Role {
  id?: { id: string };
  createdTime?: number;
  tenantId?: { id: string };
  customerId?: { id: string };
  name: string;
  type: 'GENERIC' | 'GROUP';
  permissions: { [resource: string]: string[] } | { operations: string[] };
  additionalInfo?: string;
  version?: number;
}

@Injectable({ providedIn: 'root' })
export class RoleService {

  constructor(private http: HttpClient) {}

  public saveRole(role: Role, config?: RequestConfig): Observable<Role> {
    return this.http.post<Role>('/api/role', role, defaultHttpOptionsFromConfig(config));
  }

  public getRole(id: string, config?: RequestConfig): Observable<Role> {
    return this.http.get<Role>(`/api/role/${id}`, defaultHttpOptionsFromConfig(config));
  }

  public getRoles(pageLink: PageLink, config?: RequestConfig): Observable<PageData<Role>> {
    return this.http.get<PageData<Role>>(`/api/roles${pageLink.toQuery()}`, defaultHttpOptionsFromConfig(config));
  }

  public deleteRole(id: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/role/${id}`, defaultHttpOptionsFromConfig(config));
  }

  public assignRole(userId: string, roleId: string, customerId?: string, config?: RequestConfig): Observable<void> {
    const query = customerId ? `?customerId=${customerId}` : '';
    return this.http.post<void>(`/api/user/${userId}/role/${roleId}${query}`, {}, defaultHttpOptionsFromConfig(config));
  }

  public revokeRole(userId: string, roleId: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/user/${userId}/role/${roleId}`, defaultHttpOptionsFromConfig(config));
  }

  public getUserRoles(userId: string, config?: RequestConfig): Observable<Role[]> {
    return this.http.get<Role[]>(`/api/user/${userId}/roles`, defaultHttpOptionsFromConfig(config));
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/role/RoleService.java
/**
 * Copyright © 2016-2026 The Thingsboard Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.thingsboard.server.dao.role;

import org.thingsboard.server.common.data.id.CustomerId;
import org.thingsboard.server.common.data.id.EntityId;
import org.thingsboard.server.common.data.id.RoleId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.id.UserId;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.PageLink;
import org.thingsboard.server.common.data.role.Role;

import java.util.List;
import java.util.Set;

public interface RoleService {

    Role saveRole(Role role);

    Role findById(TenantId tenantId, RoleId id);

    PageData<Role> findByTenantId(TenantId tenantId, PageLink pageLink);

    void deleteById(TenantId tenantId, RoleId id);

    // --- User <-> Role assignment ---

    void assignRole(TenantId tenantId, UserId userId, RoleId roleId, CustomerId customerId);

    void revokeRole(TenantId tenantId, UserId userId, RoleId roleId);

    List<Role> findRolesForUser(TenantId tenantId, UserId userId);

    // --- Permission lookup ---

    /**
     * Returns the set of granted operation names for the given user against the given resource. Combines
     * GENERIC-role permissions (resource → operations map) plus GROUP-role grants matching {@code entityId}.
     * Returns an empty set if the user has no role grants.
     */
    Set<String> findGrantedOperations(TenantId tenantId, UserId userId, String resourceName, EntityId entityId);
}
